components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -281,6 +288,175 @@ private void handleRequest(HttpServletRequest req, HttpServletResponse resp, boo
         }
     }
 
+    private void handleInvalidRequestMessage(HttpServletRequest req, HttpServletResponse resp, String sessionId)
+            throws IOException, IdentityException, ServletException {
+
+        log.debug("Invalid request message or single logout message ");
+
+        if (sessionId == null) {
+            String errorResp = SAMLSSOUtil.buildErrorResponse(
+                    SAMLSSOConstants.StatusCodes.REQUESTOR_ERROR,
+                    "Invalid request message", null);
+            sendNotification(errorResp, SAMLSSOConstants.Notification.INVALID_MESSAGE_STATUS,
+                    SAMLSSOConstants.Notification.INVALID_MESSAGE_MESSAGE, null, req, resp);
+        } else {
+            // Non-SAML request are assumed to be logout requests
+            sendToFrameworkForLogout(req, resp, null, null, sessionId, true,
+                    false);
+        }
+    }
+
+    private void handleSAMLResponse(HttpServletRequest req, HttpServletResponse resp, String samlResponse,
+                                    String sessionId, boolean isPost)
+            throws IdentityException, IOException, ServletException {
+
+        XMLObject response;
+
+        if (isPost) {
+            response = SAMLSSOUtil.unmarshall(SAMLSSOUtil.decodeForPost(samlResponse));
+        } else {
+            response = SAMLSSOUtil.unmarshall(SAMLSSOUtil.decode(samlResponse));
+        }
+
+        if (!(response instanceof LogoutResponseImpl)) {
+            handleInvalidRequestMessage(req, resp, sessionId);
+        } else {
+            String inResponseToId = ((LogoutResponseImpl) response).getInResponseTo();
+            FrontChannelSLOParticipantInfo frontChannelSLOParticipantInfo =
+                    getFrontChannelSLOParticipantInfo(inResponseToId);
```

Comment:
TODO: Need to validate Response signature if required.
TODO: Check currentSLOInvokedParticipant is equaled to ((LogoutResponseImpl) response).getIssuer().getValue()

If not handleInvalidRequestMessage(req, resp, sessionId);